### PR TITLE
Suggested refactor for Processing artifacts

### PR DESCRIPTION
**Pull request by sjmiles**
* refactor ImageClassifier recipes into processing and UI portions, leave selection of MlModel up to strategizer

also, lots of extra stuff:

* rearranged folders to my currently suggested best practice structure (particles in root)
* used named types instead of generic `MlResource` for various service references
  * named types allow the strategizer to do better matching
  * note that references are intended to proxy for actual objects and be referred to as if they were the target objects
* renamed some particles files: it's helps locate sources when particle-names match the file-names (there may still be some that should be adjusted)
* build tf.js high-level API for accessing tf-service, provides nicer syntax for particles
* collapsed particle sources to near-minimal loc
* invented a smart `set` method for operating on handles (experimental)

### CODE REVIEW

File: particles/Processing/Dispose.particle [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import './schemas/MlResource.schema'
+
+particle Dispose in './js/Dispose.js'
+  in * {Number ref} resource
```
Comment (alxrsngrtn) [inline particles/Processing/Dispose.particle]: I like the anonymous inline definition for the resource. This is a pattern I discovered late and wish I had used more.

File: particles/Processing/Dispose.particle [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+  //description `Clean up resources.`
```
Comment (alxrsngrtn) [inline particles/Processing/Dispose.particle]: Why don't we want a description here?

File: particles/Processing/ImageToTensor.particle [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import './schemas/Image.schema'
+import './schemas/MlResource.schema'
+
+particle ImageToTensor in './js/ImageToTensor.js'
+  in Image {(Text or URL) url} image
+  out TfTensor imageTensor
+  //description `Converts an Image URL to a Tensor3D`
```
Comment (alxrsngrtn) [inline particles/Processing/ImageToTensor.particle]: I like the flat folder structure + the idea that the name of the file matches the particle.

File: particles/Processing/Dispose.particle (hunk 2 above)
Comment (sjmiles) [reply particles/Processing/Dispose.particle]: These descriptions are for building into Suggestions for the end-user.

Comment (sjmiles) [reply particles/Processing/Dispose.particle]: Also, I forgot to put this in my notes ... the DescriptionFormatter will attempt to form an English sentence out of descriptions, so you want to avoid (1) capitals (unless for a proper-name) and (2) ending punctuation (e.g., a corrected description would be `clean up resources`.)

File: particles/Processing/LoadModel.particle [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import './schemas/MlResource.schema'
+import './schemas/MlModel.schema'
+
+particle LoadGraphModel in './js/LoadGraphModel.js'
+  in MlModel modelSpec
```
Comment (alxrsngrtn) [inline particles/Processing/LoadModel.particle]: Do you think it would be useful to rename `MlModel` to `MlModelReference` or `MlModelSpecification`? 
Further, do you think this is a good use case for inline schemas?

Comment (alxrsngrtn) [reply particles/Processing/LoadModel.particle]: I haven't seen all the PR, but I could see down-the-road that MlModel could refer to an in-memory model (or else, the actual model artifact) vs this, which is the location of the model.

File: particles/Processing/PresentLabel.particle [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import './schemas/Inference.schema'
+import './schemas/MlResource.schema'
+
+particle PresentLabel in './js/PresentLabel.js'
+  in [Classification] predictions
+  consume root
+  //description `show the top k labels (and confidence values)`
```
Comment (alxrsngrtn) [inline particles/Processing/PresentLabel.particle]: Is there a convention for writing documentation for particles, e.g. an equivalent to a javadoc? 

If so, I think we can re-use these descriptions as documentation if we don't want them here.

File: particles/Processing/LoadModel.particle (hunk 4 above)
Comment (sjmiles) [reply particles/Processing/LoadModel.particle]: It should be named as for what it represents, and I'm not sure we have the domain knowledge yet. But 'MlModelSpecification' is ok. I think 'Reference' in 'MlModelReference' is too vague.

Using inline schemas is a more complicated issue then it appears, let's discuss in person?

File: particles/Processing/PresentLabel.particle (hunk 5 above)
Comment (sjmiles) [reply particles/Processing/PresentLabel.particle]: There is no convention for that, we can consider it.

Agreed, documentation value is why I didn't erase them completely.

File: particles/Processing/TopLabels.particle [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import './schemas/MlResource.schema'
+import './schemas/MlModel.schema'
+import './schemas/Inference.schema'
+
+particle TopLabels in './js/TopLabels.js'
+  in TfYHat yHat
```
Comment (alxrsngrtn) [inline particles/Processing/TopLabels.particle]: Why is there a type for `TfYHat`? 
These should be `TfTensor`s.

File: particles/Processing/Dispose.particle (hunk 1 above)
Comment (sjmiles) [reply particles/Processing/Dispose.particle]: This particular usage is important, because there are multiple **unique** TfThing schemas. This particular pattern matches *any* schema that contains `{Number ref}` so it can match the various types.

File: particles/Processing/TopLabels.particle (hunk 6 above)
Comment (sjmiles) [reply particles/Processing/TopLabels.particle]: Ok, I simply didn't know.

Comment (alxrsngrtn) [reply particles/Processing/TopLabels.particle]: If you want to have different types to take advantage of matching, I would prefer another means, like tagging. 
This just rubs me the wrong way in terms of ML conventions. It's kinda like saying that there is a type, `Array`, and then another type `OutputArray`, where you're supposed to use the latter only for the return of functions. But really, both are arrays.

Besides just rubbing me the wrong way, this limits what operations we are able to call. I should be able to call `Normalize` or `Reshape` on the output of `Infer`, for instance.

File: particles/Processing/js/Dispose.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+defineParticle(({DomParticle, resolver, log}) => {
+
+  importScripts(resolver(`$here/tf.js`));
```
Comment (alxrsngrtn) [inline particles/Processing/js/Dispose.js]: Is it possible that we can inject tf? 

```
defineParticle(({DomParticle, resolver, log, tf}) => {
...
```

File: particles/Processing/js/ExpandDims.js [hunk 8]
```diff
@@ -8,20 +8,18 @@
...
+  importScripts(resolver(`$here/tf.js`));
 
   return class extends DomParticle {
-    update({resource}) {
-      if (resource) {
-        this.apply(resource);
+    async update({tensor, axis}) {
+      if (tensor && axis) {
+        log('Expanding Dimensions...');
+        await this.set('newTensor', await this.tf.expandDims(tensor, axis));
```
Comment (alxrsngrtn) [inline particles/Processing/js/ExpandDims.js]: This is really clean. I like it a lot!

I wasn't aware that we could just `set` handles. This makes things really clear and simple.

Comment (alxrsngrtn) [reply particles/Processing/js/ExpandDims.js]: One issue: `newTensor`, to the best of my knowedge, should be "boxed", e.g. 
```
await this.set('newTensor', {ref: await this.tf.expandDims(tensor, axis)});
```

Comment (alxrsngrtn) [reply particles/Processing/js/ExpandDims.js]: Just checked. You take care of this in the `tf.js`. Brilliant.

File: particles/Processing/js/ImageSelector.js [hunk 9]
```diff
@@ -20,12 +20,10 @@ defineParticle(({DomParticle, html, resolver, log}) => {
       <button on-click="onSubmit">Submit</button>
       <br><br>
       <img src="{{url}}">
-      <br><br>
-      <div slotid="imageView"></div>
     </div>
   `;
 
-  const defaultImage = resolver(`ImageSelector/../../../Services/assets/waltbird.jpg`);
+  const defaultImage = resolver(`https://$particles/Services/assets/waltbird.jpg`);
```
Comment (alxrsngrtn) [inline particles/Processing/js/ImageSelector.js]: :+1:

File: particles/Processing/js/ImageToTensor.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+defineParticle(({DomParticle, resolver, log}) => {
+
+  importScripts(resolver(`$here/tf.js`));
+
+  return class extends self.TfMixin(DomParticle) {
```
Comment (alxrsngrtn) [inline particles/Processing/js/ImageToTensor.js]: Why do we need to mix in Tf?

Comment (alxrsngrtn) [reply particles/Processing/js/ImageToTensor.js]: Does this construct `this.tf`, in some sense?

File: particles/Processing/js/ExpandDims.js [hunk 11]
```diff
@@ -8,20 +8,18 @@
  * http://polymer.github.io/PATENTS.txt
  */
 
-defineParticle(({DomParticle, log}) => {
+defineParticle(({DomParticle, resolver, log}) => {
+
+  importScripts(resolver(`$here/tf.js`));
 
   return class extends DomParticle {
```
Comment (alxrsngrtn) [inline particles/Processing/js/ExpandDims.js]: Why is it that in other places, we apply the TfMixin, but here we do not?

File: particles/Processing/js/LoadGraphModel.js [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+defineParticle(({DomParticle, log, resolver}) => {
+
+  importScripts(resolver(`$here/tf.js`));
+
+  return class extends self.TfMixin(DomParticle) {
+    async update({modelSpec}) {
+      if (modelSpec) {
+        const model = await this.tf.loadGraphModel(resolver(modelSpec.location), modelSpec.options);
```
Comment (alxrsngrtn) [inline particles/Processing/js/LoadGraphModel.js]: Question: Will the resolver work in test code?

Comment (alxrsngrtn) [reply particles/Processing/js/LoadGraphModel.js]: What happens if `modelSpec.options` is undefined?

File: particles/Processing/TopLabels.particle (hunk 6 above)
Comment (sjmiles) [reply particles/Processing/TopLabels.particle]: The distinct types should only be for distinct types. I just didn’t realize yHat was a Tensor.

If there are other mistakes we’ll fix them.

File: particles/Processing/js/ParseLabels.js [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+    async update({url}) {
+      if (url) {
+        log('Parsing labels file...');
+        const resolvedUrl = resolver(url.labelsUrl);
+        const response = await fetch(resolvedUrl);
+        const text = await response.text();
+        const labels = text.split('\n').map(label => ({label}));
+        await this.set('labels', labels);
+        log('Parsed.');
+
```
Comment (alxrsngrtn) [inline particles/Processing/js/ParseLabels.js]: rm space

File: particles/Processing/js/Dispose.js (hunk 7 above)
Comment (sjmiles) [reply particles/Processing/js/Dispose.js]: Yes, it is possible. But right now tf.js belongs strictly to the particles and not to runtime itself, which is a desirable property.

File: particles/Processing/js/ExpandDims.js (hunk 8 above)
Comment (sjmiles) [reply particles/Processing/js/ExpandDims.js]: ‘set’ is experimental api I put into tf.js as a try-out. See notes in that file.

File: particles/Processing/js/ResizeBilinear.js [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+defineParticle(({DomParticle, resolver, log}) => {
+
+  importScripts(resolver(`$here/tf.js`));
+
+  return class extends self.TfMixin(DomParticle) {
+    async update({images, size, options}) {
+      if (images && images.ref && size) {
```
Comment (alxrsngrtn) [inline particles/Processing/js/ResizeBilinear.js]: hmm... should we be checking that `ref` is defined elsewhere and not just here? 
What if `images.ref` is zero?

File: particles/Processing/js/TopLabels.js [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+
+defineParticle(({DomParticle, resolver, log}) => {
+
+  importScripts(resolver(`$here/tf.js`));
+
+  return class extends self.TfMixin(DomParticle) {
+    async update({yHat, labels, k}) {
+      if (yHat && labels) {
+        const topK = k || 5;
+        log(`Converting tensor output to top ${topK} labels...`);
```
Comment (alxrsngrtn) [inline particles/Processing/js/TopLabels.js]: I like the string interpolation here.

File: particles/Processing/js/tf.js [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+'use strict';
+
+self.Tf = class {
+  constructor(scope) {
+    this.scope = scope;
+  }
+  async call(name, args) {
```
Comment (alxrsngrtn) [inline particles/Processing/js/tf.js]: Nice. This makes maintaining the rest of this particle-side api cleaner.

File: particles/Processing/js/tf.js [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+  async imageToTensor({url}) {
+    return {
+      ref: await this.call('tf.imageToTensor', {imageUrl: url})
+    };
+  }
+  async resizeBilinear({ref: images}, size, options) {
+   const ref = await this.call('tf.resizeBilinear', {
+     images,
+     size: size.map(s => s.dim),
+     alignCorners: options ? options.alignCorners : false
```
Comment (alxrsngrtn) [inline particles/Processing/js/tf.js]: I see --> You're handling the `undefined` issue here.

File: particles/Processing/Dispose.particle (hunk 2 above)
Comment (alxrsngrtn) [reply particles/Processing/Dispose.particle]: Good notes, thank you.

File: particles/Processing/LoadModel.particle (hunk 4 above)
Comment (alxrsngrtn) [reply particles/Processing/LoadModel.particle]: sure thing.

File: particles/Processing/js/Dispose.js (hunk 7 above)
Comment (alxrsngrtn) [reply particles/Processing/js/Dispose.js]: Cool. We can defer this to an issue for a future PR.

Comment (alxrsngrtn) [reply particles/Processing/js/Dispose.js]: Or, have a discussion about it over there.

File: particles/Processing/js/ExpandDims.js (hunk 11 above)
Comment (alxrsngrtn) [reply particles/Processing/js/ExpandDims.js]: If you want to have access to `set`, I'm pretty sure you need to add the mixin.

File: particles/Processing/recipes/ImageClassifier.recipes [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+import '../Reshape.particle'
+import '../TopLabels.particle'
+import '../ImageSelector.particle'
+import '../PresentLabel.particle'
+import '../ModelSelector.particle'
+
+import './ClassifierStatics.manifest'
+
+recipe ImageClassifierUI
+  // since we `use`, this recipe is incomplete and coalescer is triggered
```
Comment (alxrsngrtn) [inline particles/Processing/recipes/ImageClassifier.recipes]: This is a really useful fact.

Comment (alxrsngrtn) [reply particles/Processing/recipes/ImageClassifier.recipes]: I didn't realize that this works in both directions (consuming and producing handles).

File: particles/Processing/recipes/ImageClassifier.recipes [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+    tensor <- reshapedImages
+    model <- model
+    yHat -> yHat
+
+  TopLabels
+    yHat <- yHat
+    labels <- labels
+    predictions -> predictions
+
+  description `${LoadGraphModel.model}`
```
Comment (alxrsngrtn) [inline particles/Processing/recipes/ImageClassifier.recipes]: I didn't know there was interpolation in descriptions! Can we also put js logic in side `${}`?

File: src/platform/loader-web.js [hunk 20]
```diff
@@ -35,7 +35,11 @@ export class PlatformLoader extends PlatformLoaderBase {
   // Below here invoked from inside Worker
   async loadParticleClass(spec) {
     const clazz = await this.requireParticle(spec.implFile, spec.implBlobUrl);
-    clazz.spec = spec;
+    if (clazz) {
```
Comment (alxrsngrtn) [inline src/platform/loader-web.js]: This is unrelated, right?

File: particles/Processing/js/ExpandDims.js (hunk 11 above)
Comment (sjmiles) [reply particles/Processing/js/ExpandDims.js]: Yes, thats just a mistake.

File: src/platform/loader-web.js (hunk 20 above)
Comment (sjmiles) [reply src/platform/loader-web.js]: Yes

File: particles/Processing/recipes/ImageClassifier.recipes (hunk 19 above)
Comment (sjmiles) [reply particles/Processing/recipes/ImageClassifier.recipes]: No, its very basic.

File: particles/Processing/js/ImageToTensor.js (hunk 10 above)
Comment (sjmiles) [reply particles/Processing/js/ImageToTensor.js]: Yes. I didn't like the boilerplate I needed to create the tf instance, so I decided to make it a mixin and take advantage of life-cycle methods (e.g. `constructor`). We can salt to taste.

File: particles/Processing/js/LoadGraphModel.js (hunk 12 above)
Comment (sjmiles) [reply particles/Processing/js/LoadGraphModel.js]: `resolver` should work on all platforms, if not it's a bug.

File: particles/Processing/js/ResizeBilinear.js (hunk 14 above)
Comment (sjmiles) [reply particles/Processing/js/ResizeBilinear.js]: Good catch, I didn't see this through. Checking .ref everywhere is problematic, I'd rather have an invariant that ref is never falsey. I'll make sure the reference generator is 1-based.

File: particles/Processing/js/TopLabels.js (hunk 15 above)
Comment (sjmiles) [reply particles/Processing/js/TopLabels.js]: That's good, because [you wrote it](https://github.com/PolymerLabs/arcs/blame/4b89c9eaa9041cac1ee9fb4bcaa64103fa2b1dba/particles/Processing/particles/js/TopLabels.js#L25). :)

File: src/services/reference-manager.ts [hunk 21]
```diff
@@ -10,7 +10,7 @@
 
 import {logFactory} from '../platform/log-web.js';
 
-const log = logFactory('resource-mgr');
+const log = logFactory('reference-mgr');
```
Comment (alxrsngrtn) [inline src/services/reference-manager.ts]: Can we extract this rename into a separate PR?

Comment (sjmiles) [reply src/services/reference-manager.ts]: Why?
